heft-plugins/heft-storybook-plugin/src/StorybookPlugin.ts

```diff
@@ -0,0 +1,237 @@
+   */
+  public apply(
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration,
+    options: IStorybookPluginOptions
+  ): void {
+    this._logger = heftSession.requestScopedLogger(TASK_NAME);
+
+    if (!options.storykitPackageName) {
+      throw new Error(
+        `The ${TASK_NAME} task cannot start because the "storykitPackageName"` +
+          ` plugin option was not specified`
+      );
+    }
+
+    const parseResult: IParsedPackageNameOrError = PackageName.tryParse(options.storykitPackageName);
+    if (parseResult.error) {
+      throw new Error(
+        `The ${TASK_NAME} task cannot start because the "storykitPackageName"` +
+          ` plugin option is not a valid package name: ` +
+          parseResult.error
+      );
+    }
+    this._storykitPackageName = options.storykitPackageName;
+
+    if (!options.startupModulePath) {
+      throw new Error(
+        `The ${TASK_NAME} task cannot start because the "startupModulePath"` +
+          ` plugin option was not specified`
+      );
+    }
+    this._startupModulePath = options.startupModulePath;
+
+    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
+      // TODO: Expose an API for custom CLI parameters similar to HeftSession.registerAction()
+      if (process.argv.indexOf('--storybook') < 0) {
+        this._logger.terminal.writeVerboseLine(
+          'The command line does not include "--storybook", so bundling will proceed without Storybook'
+        );
+        return;
+      }
+
+      this._logger.terminal.writeVerboseLine(
+        'The command line includes "--storybook", redirecting Webpack to Storybook'
+      );
+
+      build.hooks.preCompile.tap(PLUGIN_NAME, (preCompile: IPreCompileSubstage) => {
+        preCompile.hooks.run.tap(PLUGIN_NAME, () => {
+          this._onPreCompile(heftSession, heftConfiguration);
+        });
+      });
+
+      build.hooks.bundle.tap(PLUGIN_NAME, (bundle: IBundleSubstage) => {
+        bundle.hooks.configureWebpack.tap(
+          { name: PLUGIN_NAME, stage: Number.MAX_SAFE_INTEGER },
+          (webpackConfiguration: unknown) => {
+            // Discard Webpack's configuration to prevent Webpack from running
+            return null;
+          }
+        );
+
+        bundle.hooks.run.tapPromise(PLUGIN_NAME, async () => {
+          await this._onBundleRunAsync(heftSession, heftConfiguration);
+        });
+      });
+    });
+  }
+
+  private _onPreCompile(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    this._logger.terminal.writeVerboseLine(`Probing for "${this._storykitPackageName}"`);
+
+    // Example: "/path/to/my-project/node_modules/my-storykit"
+    let storykitFolder: string;
+    try {
+      storykitFolder = Import.resolvePackage({
+        packageName: this._storykitPackageName,
+        baseFolderPath: heftConfiguration.buildFolder
+      });
+    } catch (ex) {
+      throw new Error(`The ${TASK_NAME} task cannot start: ` + ex.message);
```

Comment:
```suggestion
      this._logger.emitError(new Error(`The ${TASK_NAME} task cannot start: ` + ex.message));
      return;
```

Reply:
IIRC `this._logger.emitError()` allows `heft start` to entire the watch loop.  That seems undesirable for a completely broken configuration.

Reply:
That's generally the approach I've taken: tell the user that things are broken, but try to continue and let Heft try to recover.

Reply:
I'm not sure I agree:  For a problem such as a syntax error in source code, yes, `heft start` should **log** the error message and then proceed normally, so it can recover if the user fixes the problem.

But for a problem such as an invalid plugin configuration, we should **throw**, because Heft does not watch for `heft.json` changes.  This is more obvious to users if `heft start` fails immediately.  We want the error message to be prominently displayed, rather than potentially scrolling off the screen behind other log output.

I don't see how Heft can distinguish these cases, so the **throw-versus-log** decision must be made by the plugin author.  We can speculate about possible future improvements for watching `heft.json`, but plugin authors can't be expected to consider that.